### ISSUE TITLE
Cancelled builds on startup with multiple SingleBranchSchedulers

### ISSUE THREAD

**Issue by LuluDavid** [buildbot/buildbot#6134]
Hi, I opened this issue because I found a problem that might be similar to #4686 and #3762.  
I am working with HgPoller but I extended it because I want to poll the initial commits as changesets to test (whereas they are considered as references not to test in the default implementation). I also created multiple SingleBranchSchedulers to listen to the polled repositories, each scheduler triggering a few builders. I also fire a Stable Timer of 30 seconds for each of the single branch scheduler.  

So when I restart my instance of buildbot, it should test all polled branches for all builders that listen to these commits. Locally, it works perfectly, and I get all expected buildrequests in the UI. But when I tried to deploy it on a distant Virtual Machine, it did not work as expected, with only the few first buildrequests being taken into account. After a lot of debugging, it seems that the missing changes are polled as expected by the HgPoller, but they are never processed by the schedulers then.  

After this initial fail, it seems then to work fine (all buildrequests happen as expected). So it just comes from a concurrency problem when all polled changes should trigger builds as for  (maybe it comes from concurrent accesses to buildbot's internal changes database). Also, the fact it works locally (16 GB of RAM, 8 CPUs)  but not on my VM (8 GB, 4 CPUs) shows that it stands out of a hardware performance problem underneath the python code, and made me think about #3762.

Thanks for reading, hope my report is exhaustive enough to track the cause of this error.

**Comment by LuluDavid** [comment]
Well the only workaround I found is to extend AnyBranchScheduler to dispatch build requests according to the branch receiving the commit, as it seems it was a concurrency matter between my multiple SingleBranchSchedulers.
So the bug is not resolved, except if some people decide to replace the current implementation of SingleBranchSchedulers with an AnyBranchScheduler that does not filter commits with a change_filter on branches, but filters them later in addBuildsetForSourceStamps.

**Comment by p12tic** [comment]
@LuluDavid Could you please connect to the database that the Buildbot uses and issues this query:

```
select complete,results,submitted_at,complete_at from buildrequests where submitted_at >= $start_time and submitted_at <= $end_time;
```

Please replace $start_time and $end_time with the UNIX time of when the nighly schedulers fired plus minus, say, 5 minutes, so that all nearby build requests are captured. https://www.epochconverter.com/ is a good resource for converting human time to UNIX time (please double-check the timezones).

**Comment by p12tic** [comment]
Also, is the `collapseRequests` (http://docs.buildbot.net/current/manual/configuration/global.html#merging-build-requests) customized in your buildbot configuration?

**Comment by p12tic** [comment]
I think this bug is fixed in #6152. It's most likely a duplicate of #4686 and/or #4642. I will thus close the issue.

If the bug is still present in Buildbot v3.3.0 once it's released, please reopen this issue.

Thank you a lot for your work investigating the bug.
